src/com/ms/silverking/cloud/dht/daemon/storage/WritableSegmentBase.java

```diff
@@ -38,419 +38,432 @@
+        //    return SegmentStorageResult.invalidVersion;
+        //} else {
+        byte[] existingChecksum;
+        byte[] newChecksum;
+
+        existingChecksum = getChecksum(existingOffset);
+        newChecksum = getChecksum(offset);
+        if (ArrayUtil.compareSigned(existingChecksum, newChecksum, ArrayUtil.MismatchedLengthMode.Ignore) == 0) {
+          return SegmentStorageResult.stored;
+        } else {
+          if (debugPut) {
+            Log.warning(
+                String.format("Checksums failed to compare: %s %s", StringUtil.byteArrayToHexString(existingChecksum),
+                    StringUtil.byteArrayToHexString(newChecksum)));
+          }
+          return SegmentStorageResult.mutation;
         }
-        if (existingOffset == CuckooBase.keyNotFound ) {
-            // no offset for the key; add the mapping
-            if (nsOptions.getVersionMode() == NamespaceVersionMode.SINGLE_VERSION || nsOptions.getStorageType() == StorageType.RAM) {
+        //}
+      } else {
+        if (existingOffset >= 0) {
+          long existingVersion;
+          long existingCreationTime;
+          long creationTime;
+
+          if (debugPut) {
+            Log.warning("single key associated: ", KeyUtil.keyToString(key));
+          }
+          // single key is associated, create an offset list
+          offsetList = offsetListStore.newOffsetList();
+          existingVersion = getVersion(existingOffset);
+
+          if (nsOptions.getRevisionMode() == RevisionMode.UNRESTRICTED_REVISIONS) {
+            existingCreationTime = getCreationTime(existingOffset);
+            creationTime = getCreationTime(offset);
+          } else {
+            existingCreationTime = 0;
+            creationTime = 0;
+          }
+
+          if (nsOptions.getRevisionMode() == RevisionMode.UNRESTRICTED_REVISIONS || version > existingVersion) {
+            offsetList.putOffset(existingVersion, existingOffset, existingCreationTime);
+            offsetList.putOffset(version, offset, creationTime);
+            if (debugPut) {
+              Log.warning("removing existing mapping: ", KeyUtil.keyToString(key));
+            }
+            boolean removed;
+            removed = keyToOffset.remove(key);
+            if (debugPut || Log.levelMet(Level.FINE)) {
+              Log.warning("removed: ", removed);
+              Log.warning("pkc.get: ", keyToOffset.get(key));
+              Log.warning("putting new mapping: ",
+                  KeyUtil.keyToString(key) + " " + -((RAMOffsetList) offsetList).getIndex());
+            }
+            try {
+              keyToOffset.put(key, -((RAMOffsetList) offsetList).getIndex());
+            } catch (TableFullException tfe) {
+              Log.warning("Segment pkc full. Creating new table");
+              keyToOffset = IntArrayCuckoo.rehashAndAdd((IntArrayCuckoo) keyToOffset, key,
+                  -((RAMOffsetList) offsetList).getIndex());
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 264:270, 195:201.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.